java-scala-diff.md

```diff
@@ -0,0 +1,285 @@
+<!--
+    Copyright (c) 2011-present Sonatype, Inc. All rights reserved.
+    Includes the third-party code listed at http://links.sonatype.com/products/clm/attributions.
+    "Sonatype" is a trademark of Sonatype, Inc.
+-->
+### Some differences between Java and Scala styles
+
+#### Wrapping long lines
+```java
+// in Java, indent 4 spaces
+stringBuilder
+    .append("one ")
+    .append("two ")
+    .append("three");
+doSomnething(
+    param1,
+    param2, 
+    param3);
+```
+```scala
+// in Scala, indent 2 spaces
+stringBuilder
+  .append("one ")
+  .append("two ")
+  .append("three")
+doSomnething(
+  param1,
+  param2, 
+  param3)
+
+// but 4 spaces for constructor parameters
+class Person(
+    name: String,
+    birthDate: Date,
+    shoeSize: Double)
+  extends Entity {
+  def firstMethod: Foo = ...
+}
+```
+```java
+// in Java, inflix operator on next line
+int x = 1 + 2 + 3 + 4 + 5 + 6 + 7 + 8 + 9
+    + 10 + 11 + 12 + 13 + 14 + 15 + 16
+    + 17 + 18 + 19 + 20;
+```
+```scala
+// in Scala, inflix method on previous line, otherwise compiler might infer semicolon unexpectedly
+val x = 1 + 2 + 3 + 4 + 5 + 6 + 7 + 8 + 9 +
+  10 + 11 + 12 + 13 + 14 + 15 + 16 + 17 +
+  18 + 19 + 20
+```
+
+#### Wrapping long parameter lists
+```scala
+// in Scala, each parameter on separate line, indented from first line (not parenthesis)
+foo(
+  someVeryLongFieldName,
+  andAnotherVeryLongFieldName,
+  "this is a string",
+  3.1415926)
+```
+
+#### Upper/lower case
+```java
+// in Java, classes, interfaces, annotations are in UpperCamelCase
+@MyAnnotation
+class MyClass implements MyInterface { ... }
+```
+```scala
+// in Scala, classes and traits are in UpperCamelCase, but annotations are in lowerCamelCase
+@myAnnotation
+class MyClass extends MyTrait { ... }
+
+// objects are usually in UpperCamelCase but can be in lowerCamelCase if they mimic a package or a function
+object ast {
+  sealed trait Expr
+
+  case class Plus(e1: Expr, e2: Expr) extends Expr { ... }
+}
+
+object inc {
+ def apply(x: Int): Int = x + 1
+}
+```
+```java
+// in Java, constants are in UPPER_CASE
+class Foo {
+  static final int MY_CONSTANT = 239;
+}
+```
+```scala
+// in Scala, constants are in UpperCamelCase
+object Foo {
+  val MyConstant = 239
+}
+```
+
+#### Getters, setters
+```java
+// in Java, “get”/“set” prefix
+class Foo {
+  int getBar() {...}
+
+  void setBar(int bar) {...}
+}
+```
+```scala
+// in Scala, just the name of a property
+class Foo {
+  def bar: Int = ...
+
+  def bar_=(bar: Int) { ... }
+}
+
+// if getter/setter are backed by a field, underscore prefix to disambiguate
```

Comment:
should we make any mention that we'd only back it by a field when necessary? Meaning, we don't want someone reading this and thinking that's the right way to do it all the time.